### PR TITLE
new cumprod atom

### PR DESCRIPTION
**Pull request by Transurgeon**
## Description
Please include a short summary of the change.
This PR adds a new cvxpy atom that is analog to [numpy.cumprod](https://numpy.org/doc/stable/reference/generated/numpy.cumprod.html).
Issue link (if applicable): from Discord.

Note for future-self:
The implementation ended up being just calling cumsum because after the dgp2dcp canonicalization, the atoms are actually equivalent. (in other words, cumprod is the equivalent of cumsum in log-log land). 
Another way to view this is that cumsum has constraints of the following form: ``Y[:1] = X[1:] - X[:1]``, and cumprod would simply need to replace the negation operator with a division. However, the log-log-reduction actually does just that! Under the hood, it turns all divisions into negations. What a beautiful concept.
## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [x] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [x] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: cvxpy/atoms/affine/cumprod.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+
+class cumprod(AffAtom, AxisAtom):
+    """
+    Cumulative product of the elements of an expression.
+
+    Attributes
+    ----------
+    expr : CVXPY expression
+        The expression being summed.
```
Comment (PTNobel) [inline cvxpy/atoms/affine/cumprod.py:34]:
```suggestion
        The expression being multiplied.
```

File: cvxpy/atoms/affine/cumprod.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    axis : int
+        The axis to sum across if 2D.
```
Comment (PTNobel) [inline cvxpy/atoms/affine/cumprod.py:36]:
```suggestion
        The axis to multiply across.
```

**Comment by rileyjmurray** [comment]
Small note: cumprod isn't an affine atom.

File: cvxpy/atoms/affine/cumprod.py [hunk 3]
```diff
@@ -87,4 +89,6 @@ def graph_implementation(
         tuple
             (LinOp for objective, list of constraints)
         """
-        return ()
+        n = arg_objs[0]
+        A = lu.create_const(np.triu(np.ones(shape)), shape=shape)
+        return (gmatmul(A, n), [])
```
Comment (SteveDiamond) [inline cvxpy/atoms/affine/cumprod.py:94]: You can't do cumprod with a graph_implementation. You need to implement a reduction in ``cvxpy/reductions/dgp2dcp/canonicalizers/``

Comment (Transurgeon) [reply cvxpy/atoms/affine/cumprod.py:94]: oh I see, thanks a lot. I was really stuck on this and didn't know why it wasn't working.

File: cvxpy/reductions/dgp2dcp/canonicalizers/cumprod_canon.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+def cumprod_canon(expr, args):
+    out = 0.0
+    for x_i, p_i in zip(args[0], expr.p):
```
[the remark this answers is not in the document]
Comment (Transurgeon) [reply cvxpy/reductions/dgp2dcp/canonicalizers/cumprod_canon.py:18]: ok sounds good. I wasn't sure for the axis argument too, I'll think about it.

[the remark this answers is not in the document]
Comment (Transurgeon) [reply cvxpy/reductions/dgp2dcp/canonicalizers/cumprod_canon.py]: I think initial tests seems to pass by just using cumsum in the dgp2dcp canon for cumprod.

**Review by SteveDiamond — APPROVED** [review]
lgtm

**Pull request merged**